xcp_d/utils/utils.py

```diff
@@ -461,3 +472,93 @@ def estimate_brain_radius(mask_file, head_radius="auto"):
         brain_radius = head_radius
 
     return brain_radius
+
+
+def _denoise_with_nilearn(
+    raw_data,
+    confounds_file,
+    censoring_file,
+    lowpass,
+    highpass,
+    filter_order,
+    TR,
+):
+    """Denoise an array with Nilearn.
+
+    This step does the following.
+    Linearly detrend, but don't mean-center, the BOLD data.
+    Regress out confounds from BOLD data.
+    Use list of outliers to censor BOLD data during regression.
+    Temporally filter BOLD data.
+    """
+    import pandas as pd
+    from nilearn import signal
+
+    n_volumes, n_voxels = raw_data.shape
+    confounds_df = pd.read_table(confounds_file)
+
+    censoring_df = pd.read_table(censoring_file)
+    sample_mask = ~censoring_df["framewise_displacement"].to_numpy().astype(bool)
+
+    # Per xcp_d's style, censor the data first
+    raw_data_censored = raw_data[sample_mask, :]
+    confounds = confounds_df.to_numpy()
+    confounds_censored = confounds[sample_mask, :]
+    confounds_df_censored = pd.DataFrame(data=confounds_censored, columns=confounds_df.columns)
+
+    # Orthogonalize censored nuisance regressors w.r.t. any censored signal regressors
+    signal_columns = [c for c in confounds_df.columns if c.startswith("signal__")]
+    if signal_columns:
+        warnings.warn(
+            "Signal columns detected. "
+            "Orthogonalizing nuisance columns w.r.t. the following signal columns: "
+            f"{', '.join(signal_columns)}"
+        )
+        noise_columns = [c for c in confounds_df_censored.columns if not c.startswith("signal__")]
+        temp_confounds_df = confounds_df_censored[noise_columns].copy()
+        signal_regressors = confounds_df_censored[signal_columns].to_numpy()
+        noise_regressors = temp_confounds_df.to_numpy()
+        betas = np.linalg.lstsq(signal_regressors, noise_regressors, rcond=None)[0]
+        pred_noise_regressors = np.dot(signal_regressors, betas)
+        orth_noise_regressors = noise_regressors - pred_noise_regressors
+        temp_confounds_df.loc[:, :] = orth_noise_regressors
+        confounds_df_censored = temp_confounds_df
+
+    # Then detrend and regress
+    clean_data_censored = signal.clean(
+        signals=raw_data_censored,
+        detrend=True,
+        standardize=False,
+        sample_mask=None,
+        confounds=confounds_df_censored,
+        standardize_confounds=True,
+        filter=None,
+        t_r=TR,
+        ensure_finite=True,
+    )
+
+    # Now interpolate with cubic spline interpolation
+    clean_data_interp = np.zeros((n_volumes, n_voxels), dtype=clean_data_censored.dtype)
+    clean_data_interp[sample_mask, :] = clean_data_censored
+    clean_data_interp = signal._interpolate_volumes(
+        clean_data_interp,
+        sample_mask=sample_mask,
+        t_r=TR,
+    )
```

Comment:
Switching from linear interpolation to cubic spline interpolation. Is this a problem?